lib/base/json.cpp

```diff
@@ -1,23 +1,228 @@
+ *
+ * @param value The value to be JSON serialized.
+ */
+void JsonEncoder::EncodeImpl(const Value& value)
+{
+	switch (value.GetType()) {
+		case ValueEmpty:
+			Write("null");
+			return;
+		case ValueBoolean:
+			if (value.ToBool()) {
+				Write("true");
+			} else {
+				Write("false");
+			}
+			return;
+		case ValueString:
+			EncodeValidatedJson(Utility::ValidateUTF8(value.Get<String>()));
+			return;
+		case ValueNumber:
+			if (auto ll(static_cast<long long>(value)); ll == value) {
+				EncodeValidatedJson(ll);
+			} else {
+				EncodeValidatedJson(value.Get<double>());
+			}
+			return;
+		case ValueObject: {
+			const Object::Ptr& obj = value.Get<Object::Ptr>();
+			const auto& type = obj->GetReflectionType();
+			if (type == Namespace::TypeInstance) {
+				static constexpr auto extractor = [](const NamespaceValue& v) -> const Value& { return v.Val; };
+				EncodeObject(static_pointer_cast<Namespace>(obj), extractor);
+			} else if (type == Dictionary::TypeInstance) {
+				static constexpr auto extractor = [](const Value& v) -> const Value& { return v; };
+				EncodeObject(static_pointer_cast<Dictionary>(obj), extractor);
+			} else if (type == Array::TypeInstance) {
+				auto arr(static_pointer_cast<Array>(obj));
+				EncodeImpl(BeginArray);
+				ObjectLock olock(arr);
+				for (const auto& item : arr) {
+					EncodeItem(item);
+				}
+				EncodeImpl(EndArray);
+			} else if (auto gen(dynamic_pointer_cast<ValueGenerator>(obj)); gen) {
+				EncodeImpl(BeginArray);
+				while (true) {
+					auto result = gen->Next();
+					if (!result) {
+						break; // Stop when the generator is exhausted.
+					}
+					EncodeItem(*result);
+				}
+				EncodeImpl(EndArray);
+			} else {
+				// Some other non-serializable object type!
+				EncodeValidatedJson(Utility::ValidateUTF8(obj->ToString()));
+			}
+			return;
+		}
+		default:
+			VERIFY(!"Invalid variant type.");
+	}
+}
+
+/**
+ * Encodes an item into the currently open JSON array.
+ *
+ * This is used to incrementally build a JSON array by adding items to it one by one.
+ * Calling this function outside an array context will result in an exception being thrown.
+ *
+ * @param item The item to be encoded into the JSON array.
+ */
+void JsonEncoder::EncodeItem(const Value& item)
+{
+	if (m_CtxStack.empty() || m_CtxStack.top().type != CtxType::Array) {
+		BOOST_THROW_EXCEPTION(std::logic_error("Cannot encode item outside an array context."));
+	}
+
+	writeSeparatorsAndIndentationIfNeeded();
+	EncodeImpl(item);
+	incrementCtxElements();
+}
+
+void JsonEncoder::writeSeparatorsAndIndentationIfNeeded()
+{
+	if (!m_CtxStack.empty() && m_CtxStack.top().count > 0) {
+		EncodeImpl(Comma);
+	}
+	if (m_Pretty) {
+		// If we've just started a new Object or Array and pretty printing is enabled, we'll need to
+		// write a newline and indentation before its first element is written to ensure proper formatting.
+		if (!m_CtxStack.empty() && m_CtxStack.top().count == 0) {
+			EncodeImpl(NewLine);
+		}
```

Comment:
```suggestion
		EncodeImpl(NewLine);
```

I think you can remove this condition...